<a href="https://colab.research.google.com/github/JraHike76/DS-Unit-1-Sprint-1-Data-Wrangling-and-Storytelling/blob/master/module2-make-features/%20LS_DS_112_Make_Features.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Lambda School Data Science

*Unit 1, Sprint 1, Module 2*

---

# Make Features 

- Student should be able to understand the purpose of feature engineering
- Student should be able to work with strings in pandas
- Student should be able to work with dates and times in pandas
- Student should be able to modify or create columns of a dataframe using the `.apply()` function


Helpful Links:
- [Minimally Sufficient Pandas](https://medium.com/dunder-data/minimally-sufficient-pandas-a8e67f2a2428)
- [Feature Engineering](https://en.wikipedia.org/wiki/Feature_engineering)
- Python Data Science Handbook
  - [Chapter 3.10](https://jakevdp.github.io/PythonDataScienceHandbook/03.10-working-with-strings.html), Vectorized String Operations
  - [Chapter 3.11](https://jakevdp.github.io/PythonDataScienceHandbook/03.11-working-with-time-series.html), Working with Time Series
- [Lambda Learning Method for DS - By Ryan Herr](https://docs.google.com/document/d/1ubOw9B3Hfip27hF2ZFnW3a3z9xAgrUDRReOEo-FHCVs/edit?usp=sharing)

# [Objective](#feature-engineering) - The Purpose of Feature Engineering



## Overview

Feature Engineering is the process of using a combination of domain knowledge, creativity and the pre-existing columns of a dataset to create completely new columns.

 Machine Learning models try to detect patterns in the data and then associate those patterns with certain predictions. The hope is that by creating new columns on our dataset that we can expose our model to new patterns in the data so that it can make better and better predictions.

This is largely a matter of understanding how to work with individual columns of a dataframe with Pandas --which is what we'll be practicing today!

## Follow Along

Columns of a dataframe hold each hold a specific type of data. Lets inspect some of the common datatypes found in datasets and then we'll make a new feature on a dataset using pre-existing columns.

In [0]:
import pandas as pd



In [0]:
#pandas display options
pd.set_option('display.max_rows',150)
pd.set_option('display.max_columns',100)

In [0]:
# Lets take a look at the Ames Iowa Housing Dataset:
df = pd.read_csv('https://raw.githubusercontent.com/ryanleeallred/datasets/master/Ames%20Housing%20Data/train.csv')
print(df.shape)
df.head()

(1460, 81)


Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities LotConfig LandSlope Neighborhood Condition1  \
0         Lvl    AllPub    Inside       Gtl      CollgCr       Norm   
1         Lvl    AllPub       FR2       Gtl      Veenker      Feedr   
2         Lvl    AllPub    Inside       Gtl      CollgCr       Norm   
3         Lvl    AllPub    Corner       Gtl      Crawfor       Norm   
4         Lvl    AllPub       FR2       Gtl      NoRidge       Norm   

  Condition2 BldgType HouseStyle  OverallQual  OverallCond  YearBuilt  \
0       Norm     1Fam     2Story            7            5       2003   
1       Norm     1Fam     1Story            6            8       1976   
2       Norm     1Fam     2Story            7            5       2001   
3       Norm     1Fam     2Story            7            5       1915   
4       Norm     1Fam     2Story            8            5       2000   

   YearRemodAdd RoofStyle RoofMatl Exterior1st Exterior2nd MasVnrType  \
0          2003     Gable  CompShg     VinylSd     VinylSd    BrkFace   
1          1976     Gable  CompShg     MetalSd     MetalSd       None   
2          2002     Gable  CompShg     VinylSd     VinylSd    BrkFace   
3          1970     Gable  CompShg     Wd Sdng     Wd Shng       None   
4          2000     Gable  CompShg     VinylSd     VinylSd    BrkFace   

   MasVnrArea ExterQual ExterCond Foundation BsmtQual BsmtCond BsmtExposure  \
0       196.0        Gd        TA      PConc       Gd       TA           No   
1         0.0        TA        TA     CBlock       Gd       TA           Gd   
2       162.0        Gd        TA      PConc       Gd       TA           Mn   
3         0.0        TA        TA     BrkTil       TA       Gd           No   
4       350.0        Gd        TA      PConc       Gd       TA           Av   

  BsmtFinType1  BsmtFinSF1 BsmtFinType2  BsmtFinSF2  BsmtUnfSF  TotalBsmtSF  \
0          GLQ         706          Unf           0        150          856   
1          ALQ         978          Unf           0        284         1262   
2          GLQ         486          Unf           0        434          920   
3          ALQ         216          Unf           0        540          756   
4          GLQ         655          Unf           0        490         1145   

  Heating HeatingQC CentralAir Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  \
0    GasA        Ex          Y      SBrkr       856       854             0   
1    GasA        Ex          Y      SBrkr      1262         0             0   
2    GasA        Ex          Y      SBrkr       920       866             0   
3    GasA        Gd          Y      SBrkr       961       756             0   
4    GasA        Ex          Y      SBrkr      1145      1053             0   

   GrLivArea  BsmtFullBath  BsmtHalfBath  FullBath  HalfBath  BedroomAbvGr  \
0       1710             1             0         2         1             3   
1       1262             0             1         2         0             3   
2       1786             1             0         2         1             3   
3       1717             1             0         1         0             3   
4       2198             1             0         2         1             4   

   KitchenAbvGr KitchenQual  TotRmsAbvGrd Functional  Fireplaces FireplaceQu  \
0             1          Gd             8        Typ           0         NaN   
1             1          TA             6        Typ           1          TA   
2             1          Gd             6        Typ           1          TA   
3             1          Gd            

### Specific Columns hold specific kinds of data

In [0]:
df.dtypes


Id                 int64
MSSubClass         int64
MSZoning          object
LotFrontage      float64
LotArea            int64
Street            object
Alley             object
LotShape          object
LandContour       object
Utilities         object
LotConfig         object
LandSlope         object
Neighborhood      object
Condition1        object
Condition2        object
BldgType          object
HouseStyle        object
OverallQual        int64
OverallCond        int64
YearBuilt          int64
YearRemodAdd       int64
RoofStyle         object
RoofMatl          object
Exterior1st       object
Exterior2nd       object
MasVnrType        object
MasVnrArea       float64
ExterQual         object
ExterCond         object
Foundation        object
BsmtQual          object
BsmtCond          object
BsmtExposure      object
BsmtFinType1      object
BsmtFinSF1         int64
BsmtFinType2      object
BsmtFinSF2         int64
BsmtUnfSF          int64
TotalBsmtSF        int64
Heating           object


Some columns hold integer values like the `BedroomAbvGr` which stands for "Bedrooms Above Grade." This is the number of non-basement bedrooms in the home.

For more information on specific column meanings view the [data dictionary](https://github.com/ryanleeallred/datasets/blob/master/Ames%20Housing%20Data/data_description.txt).

In [0]:
# Look at the first ten rows of the `BedroomAbvGr` column.
# Looks like integers to me!

df['BedroomAbvGr']

0       3
1       3
2       3
3       3
4       4
       ..
1455    3
1456    3
1457    4
1458    2
1459    3
Name: BedroomAbvGr, Length: 1460, dtype: int64

Some columns hold float values like the `LotFrontage` column.

In [0]:
# Look at the first ten rows of the `LotFrontage` column
df['LotFrontage']

0       65.0
1       80.0
2       68.0
3       60.0
4       84.0
        ... 
1455    62.0
1456    85.0
1457    66.0
1458    68.0
1459    75.0
Name: LotFrontage, Length: 1460, dtype: float64

Hmmm, do the values above look like floats to you?

They all have .0 on them so technically they're being stored as floats, but *should* they be stored as floats?

Lets see what all of the possible values for this column are.

In [0]:
df['LotFrontage'].value_counts()

60.0     143
70.0      70
80.0      69
50.0      57
75.0      53
65.0      44
85.0      40
78.0      25
21.0      23
90.0      23
68.0      19
24.0      19
64.0      19
73.0      18
72.0      17
79.0      17
63.0      17
55.0      17
100.0     16
51.0      15
66.0      15
74.0      15
52.0      14
59.0      13
71.0      12
82.0      12
40.0      12
43.0      12
67.0      12
57.0      12
69.0      11
76.0      11
86.0      10
53.0      10
34.0      10
92.0      10
88.0      10
62.0       9
44.0       9
84.0       9
77.0       9
35.0       9
93.0       8
61.0       8
98.0       8
96.0       8
107.0      7
120.0      7
95.0       7
58.0       7
41.0       6
30.0       6
81.0       6
89.0       6
105.0      6
48.0       6
110.0      6
94.0       6
54.0       6
91.0       6
36.0       6
47.0       5
37.0       5
83.0       5
56.0       5
87.0       5
32.0       5
42.0       4
49.0       4
102.0      4
103.0      3
108.0      3
99.0       3
104.0      3
45.0       3
130.0      3
174.0      2

In [0]:
df['MSZoning'].value_counts()

RL         1151
RM          218
FV           65
RH           16
C (all)      10
Name: MSZoning, dtype: int64

Looks to me like the `LotFrontage` column originally held integer values but was cast to a `float` meaning that each original integer values was converted to its corresponding float representation. 

Any guesses as to why that would have happened?


HINT: What's the most common `LotFrontage` value for this column?

In [0]:
# NaN is the most common value in this column. What is a NaN
import numpy as np
type(np.NaN)

float

In [0]:
df['LotFrontage'].value_counts(dropna=False)

NaN      259
60.0     143
70.0      70
80.0      69
50.0      57
75.0      53
65.0      44
85.0      40
78.0      25
21.0      23
90.0      23
68.0      19
24.0      19
64.0      19
73.0      18
72.0      17
79.0      17
63.0      17
55.0      17
100.0     16
51.0      15
66.0      15
74.0      15
52.0      14
59.0      13
71.0      12
67.0      12
57.0      12
82.0      12
43.0      12
40.0      12
76.0      11
69.0      11
53.0      10
92.0      10
88.0      10
34.0      10
86.0      10
77.0       9
35.0       9
44.0       9
84.0       9
62.0       9
93.0       8
61.0       8
98.0       8
96.0       8
107.0      7
95.0       7
120.0      7
58.0       7
41.0       6
105.0      6
54.0       6
94.0       6
48.0       6
110.0      6
30.0       6
89.0       6
81.0       6
91.0       6
36.0       6
47.0       5
83.0       5
37.0       5
87.0       5
32.0       5
56.0       5
102.0      4
42.0       4
49.0       4
99.0       3
45.0       3
104.0      3
130.0      3
103.0      3
108.0      3

`NaN` stands stands for "Not a Number" and is the default missing value indicator with Pandas. This means there were cells in this column that didn't have a LotFrontage value recorded for those homes. 

This is where domain knowledge starts to come in. Think about the context we're working with here: houses. What might a null or blank cell representing "Linear feet of street connected to property" mean in the context of a housing dataset?

Ok, so maybe it makes seanse to have some NaNs in this column. What is the datatype of a NaN value?

Perhaps some of this data is truly missing or unrecorded data, but sometimes `NaNs` are more likely to indicate something that was "NA" or "Not Applicable" to a particular observation. There could be multiple reasons why there was no value recorded for a particular feature.

Remember, that Pandas tries to maintain a single datatype for all values in a column, and therefore...

In [0]:
import numpy as np

# What is the datatype of NaN?

type(np.NaN)

float

The datatype of a NaN is float!  This means that if we have a column of integer values, but the column has even a single `NaN` that column will not be treated with the integer datatype but all of the integers will be converted to floats in order to try and preserve the same datatype throughout the entire column.

You can see already how understanding column datatypes is crucial to understanding how Pandas help us manage our data.

### Making new Features

Lets slim down the dataset and consider just a few specific columns:

- `TotalBsmtSF`
- `1stFlrSF`
- `2ndFlrSF`
- `SalePrice1`


In [0]:
# I can make a smaller dataframe with a few specific column headers
# by passing a list of column headers inside of the square brackets
df_small = df[['TotalBsmtSF', '1stFlrSF', '2ndFlrSF', 'SalePrice']].copy()
df_small.head()

TotalBsmtSF  1stFlrSF  2ndFlrSF  SalePrice
0          856       856       854     208500
1         1262      1262         0     181500
2          920       920       866     223500
3          756       961       756     140000
4         1145      1145      1053     250000

In [0]:
df_small.dtypes

TotalBsmtSF             int64
1stFlrSF                int64
2ndFlrSF                int64
SalePrice               int64
TotalSquareFootage      int64
PricePerSqFt          float64
dtype: object

### Syntax for creating new columns

When making a new column on a dataframe, we have to use the square bracket syntax of accessing a column. We can't use "dot syntax" here.

In [0]:
# Lets add up all of the square footage to get a single square footage 
# column for the entire dataset

# Using bracket syntax to make a new 'TotalSquareFootage' column

df_small['TotalSquareFootage'] = df_small['TotalBsmtSF'] + df_small['1stFlrSF'] + df_small['2ndFlrSF']
df_small.head()

TotalBsmtSF  1stFlrSF  2ndFlrSF  SalePrice  TotalSquareFootage
0          856       856       854     208500                2566
1         1262      1262         0     181500                2524
2          920       920       866     223500                2706
3          756       961       756     140000                2473
4         1145      1145      1053     250000                3343

In [0]:
# Lets make a nother new column that is 'PricePerSqFt' by
# dividing the price by the square footage
df_small['PricePerSqFt'] = df_small['SalePrice'] / df_small['TotalSquareFootage']
df_small.head()

TotalBsmtSF  1stFlrSF  2ndFlrSF  SalePrice  TotalSquareFootage  \
0          856       856       854     208500                2566   
1         1262      1262         0     181500                2524   
2          920       920       866     223500                2706   
3          756       961       756     140000                2473   
4         1145      1145      1053     250000                3343   

   PricePerSqFt  
0     81.254871  
1     71.909667  
2     82.594235  
3     56.611403  
4     74.783129

Ok, we have made two new columns on our small dataset.

- What does a **high** `PricePerSqFt` say about a home that the square footage and price alone don't capture as directly?

- What does a **low** `PricePerSqFt` say about a home that the square footage and price alone don't directly capture?



## Challenge

I hope you can see how we have used existing columns to create a new column on a dataset that say something new about our unit of observation. This is what making new features (columns) on a dataset is all about and why it's so essential to data science --particularly predictive modeling "Machine Learning." 

We'll spend the rest of the lecture and assignment today trying to get as good as we can at manipulating (cleaning) and creating new columns on datasets.

# [Objective](#work-with-strings) Work with Strings with Pandas

## Overview

So far we have worked with numeric datatypes (ints and floats) but we haven't worked with any columns containing string values. We can't simply use arithmetic to manipulate string values, so we'll need to learn some more techniques in order to work with this datatype.

## Follow Along

We're going to import a new dataset here to work with. This dataset is from LendingClub and holds information about loans issued in Q4 of 2018. This dataset is a bit messy so it will give us plenty of opportunities to clean up existing columns as well as create new ones.

The `!wget` shell command being used here does exactly the same thing that your browser does when you type a URL in the address. It makes a request or "gets" the file at that address. However, in our case the file isn't a webpage, it's a compressed CSV file. 

Try copying and pasting the URL from below into your browser, did it start an automatic download? Any URLs like this that start automatic downloads when navigated to can be used along with the `!wget` command to bring files directly into your notebook's memory.

### Load a new dataset

In [0]:
!wget https://resources.lendingclub.com/LoanStats_2018Q4.csv.zip

--2020-05-05 18:48:58--  https://resources.lendingclub.com/LoanStats_2018Q4.csv.zip
Resolving resources.lendingclub.com (resources.lendingclub.com)... 52.37.66.6, 52.25.23.92, 54.71.0.232
Connecting to resources.lendingclub.com (resources.lendingclub.com)|52.37.66.6|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: unspecified [application/zip]
Saving to: ‘LoanStats_2018Q4.csv.zip.2’

LoanStats_2018Q4.cs     [          <=>       ]  22.04M  1.94MB/s    in 12s     

2020-05-05 18:49:11 (1.84 MB/s) - ‘LoanStats_2018Q4.csv.zip.2’ saved [23108433]



We need to use the `!unzip` command to extract the csv from the zipped folder.

In [0]:
!unzip LoanStats_2018Q4.csv

Archive:  LoanStats_2018Q4.csv.zip
  inflating: LoanStats_2018Q4.csv    


We can also use bash/shell commands to look at the raw file using the `!head` and `!tail` commands

In [0]:
!head LoanStats_2018Q4.csv

Notes offered by Prospectus (https://www.lendingclub.com/info/prospectus.action)
"id","member_id","loan_amnt","funded_amnt","funded_amnt_inv","term","int_rate","installment","grade","sub_grade","emp_title","emp_length","home_ownership","annual_inc","verification_status","issue_d","loan_status","pymnt_plan","url","desc","purpose","title","zip_code","addr_state","dti","delinq_2yrs","earliest_cr_line","inq_last_6mths","mths_since_last_delinq","mths_since_last_record","open_acc","pub_rec","revol_bal","revol_util","total_acc","initial_list_status","out_prncp","out_prncp_inv","total_pymnt","total_pymnt_inv","total_rec_prncp","total_rec_int","total_rec_late_fee","recoveries","collection_recovery_fee","last_pymnt_d","last_pymnt_amnt","next_pymnt_d","last_credit_pull_d","collections_12_mths_ex_med","mths_since_last_major_derog","policy_code","application_type","annual_inc_joint","dti_joint","verification_status_joint","acc_now_delinq","tot_coll_amt","tot_cur_bal","open_acc_6m","open_act_il","op

In [0]:
!tail LoanStats_2018Q4.csv

"","","5600","5600","5600"," 36 months"," 13.56%","190.21","C","C1","","n/a","RENT","15600","Not Verified","Oct-2018","Current","n","","","credit_card","Credit card refinancing","836xx","ID","15.31","0","Aug-2012","0","","97","9","1","5996","34.5%","11","w","3082.01","3082.01","3413.23","3413.23","2517.99","895.24","0.0","0.0","0.0","Apr-2020","190.21","May-2020","Apr-2020","0","","1","Individual","","","","0","0","5996","0","0","0","1","20","0","","0","2","3017","35","17400","1","0","0","3","750","4689","45.5","0","0","20","73","13","13","0","13","","20","","0","3","5","4","4","1","9","10","5","9","0","0","0","0","100","25","1","0","17400","5996","8600","0","","","","","","","","","","","","N","","","","","","","","","","","","","","","N","","","","","",""
"","","23000","23000","23000"," 36 months"," 15.02%","797.53","C","C3","Tax Consultant","10+ years","MORTGAGE","75000","Source Verified","Oct-2018","Charged Off","n","","","debt_consolidation","Debt consolidation","352xx","AL","20.9

As we look at the raw file itself, do you see anything that might cause us trouble as we read in the CSV file to a dataframe?

In [0]:
# Read in the CSV
df = pd.read_csv('LoanStats_2018Q4.csv')

print(df.shape)
df.head()

/usr/local/lib/python3.6/dist-packages/IPython/core/interactiveshell.py:2718: DtypeWarning: Columns (0,1,2,3,4,7,13,18,19,24,25,27,28,29,30,31,32,34,36,37,38,39,40,41,42,43,44,46,49,50,51,53,54,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94,95,96,97,98,99,100,101,102,103,104,105,106,107,108,109,110,111,113,114,115,116,117,118,119,120,121,126,127,131,132,134,135,136,141,142,143) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


(128415, 1)


Notes offered by Prospectus (https://www.lendingclub.com/info/prospectus.action)
id  member_id loan_amnt funded_amnt funded_amnt_inv term       int_rate installment grade sub_grade emp_title         emp_length home_ownership annual_inc verification_status issue_d  loan_status pymnt_plan url desc purpose            title                   zip_code addr_state dti   delinq_2yrs earliest_cr_line inq_last_6mths mths_since_last_delinq mths_since_last_record open_acc pub_rec revol_bal revol_util total_acc initial_list_status out_prncp out_prncp_inv total_pymnt      total_pymnt_inv total_rec_prncp total_rec_int total_rec_late_fee recoveries collection_recovery_fee last_pymnt_d last_pymnt_amnt next_pymnt_d last_credit_pull_d collections_12_mths_ex_med mths_since_last_major_derog policy_code application_type annual_inc_joint dti_joint verification_status_joint acc_now_delinq tot_coll_amt tot_cur_bal open_acc_6m open_act_il open_il_12m open_il_24m mths_since_rcnt_il total_bal_il il_util open_rv_12m open_rv_24m max_bal_bc all_util total_rev_hi_lim inq_fi total_cu_tl inq_last_12m acc_open_past_24mths avg_cur_bal bc_open_to_buy bc_util chargeoff_within_12_mths delinq_amnt mo_sin_old_il_acct mo_sin_old_rev_tl_op mo_sin_rcnt_rev_tl_op mo_sin_rcnt_tl mort_acc mths_since_recent_bc mths_since_recent_bc_dlq mths_since_recent_inq mths_since_recent_revol_delinq num_accts_ever_120_pd num_actv_bc_tl num_actv_rev_tl num_bc_sats num_bc_tl num_il_tl num_op_rev_tl num_rev_accts num_rev_tl_bal_gt_0 num_sats num_tl_120dpd_2m num_tl_30dpd num_tl_90g_dpd_24m num_tl_op_past_12m pct_tl_nvr_dlq percent_bc_gt_75 pub_rec_bankruptcies tax_liens tot_hi_cred_lim total_bal_ex_mort total_bc_limit total_il_high_credit_limit revol_bal_joint sec_app_earliest_cr_line sec_app_inq_last_6mths sec_app_mort_acc sec_app_open_acc sec_app_revol_util sec_app_open_act_il sec_app_num_rev_accts sec_app_chargeoff_within_12_mths sec_app_collections_12_mths_ex_med sec_app_mths_since_last_major_derog hardship_flag hardship_type hardship_reason hardship_status deferral_term hardship_amount hardship_start_date hardship_end_date payment_plan_start_date hardship_length hardship_dpd hardship_loan_status orig_projected_additional_accrued_interest hardship_payoff_balance_amount hardship_last_payment_amount debt_settlement_flag debt_settlement_flag_date settlement_status settlement_date settlement_amount settlement_percentage                                    settlement_term                              
NaN NaN       35000     35000       35000            36 months  14.47%  1204.23     C     C2        Staff Physician   8 years    MORTGAGE       360000     Verified            Dec-2018 Fully Paid  n          NaN NaN  credit_card        Credit card refinancing 336xx    FL         19.9  0           Apr-1995         1              NaN                    NaN                    24       0       57259     43.2%      51        w                   0.00      0.00          38187.0468373662 38187.05        35000.00        3187.05       0.0                0.0        0.0                     Aug-2019     29882.16        NaN          Sep-2019           0                          NaN                         1           Individual       NaN              NaN       NaN                       0              0            828060      0           6           1           2           9                  112027       29      4           8           19118      36       132400           3      2           5            11                   34503       67341          46      0                        0           173                284                  8                     8              3        8                    NaN                      0                     NaN                            0                     11             11              13          20        23        16            24            11                  24       0                0            0                  5                  100            30.8         

The extra rows at the top and bottom of the file have done two things:

1) The top row has made it so that the entire dataset is being interpreted as column headers

2) The bottom two rows have been read into the 'id' column and are causing every column to have at least two `NaN` values in it.

In [0]:
# We can fix the header problem by using the 'skiprows' parameter
df = pd.read_csv('LoanStats_2018Q4.csv', header=1)

print(df.shape)
df.head()

/usr/local/lib/python3.6/dist-packages/IPython/core/interactiveshell.py:2718: DtypeWarning: Columns (0) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


(128414, 144)


id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  NaN        NaN    35000.0      35000.0          35000.0   36 months   
1  NaN        NaN     5000.0       5000.0           5000.0   36 months   
2  NaN        NaN    10000.0      10000.0          10000.0   60 months   
3  NaN        NaN    17100.0      17100.0          17100.0   36 months   
4  NaN        NaN     4000.0       4000.0           4000.0   36 months   

  int_rate  installment grade sub_grade             emp_title emp_length  \
0   14.47%      1204.23     C        C2       Staff Physician    8 years   
1   22.35%       191.86     D        D5     Director of Sales  10+ years   
2   23.40%       284.21     E        E1                   NaN   < 1 year   
3   18.94%       626.30     D        D2         Receptionist   10+ years   
4   10.72%       130.43     B        B2  Extrusion assistant   10+ years   

  home_ownership  annual_inc verification_status   issue_d loan_status  \
0       MORTGAGE    360000.0            Verified  Dec-2018  Fully Paid   
1       MORTGAGE     72000.0     Source Verified  Dec-2018  Fully Paid   
2           RENT     55000.0     Source Verified  Dec-2018     Current   
3           RENT     38000.0            Verified  Dec-2018     Current   
4       MORTGAGE     56000.0        Not Verified  Dec-2018     Current   

  pymnt_plan  url  desc             purpose                    title zip_code  \
0          n  NaN   NaN         credit_card  Credit card refinancing    336xx   
1          n  NaN   NaN  debt_consolidation       Debt consolidation    333xx   
2          n  NaN   NaN  debt_consolidation       Debt consolidation    902xx   
3          n  NaN   NaN  debt_consolidation       Debt consolidation    150xx   
4          n  NaN   NaN         credit_card  Credit card refinancing    301xx   

  addr_state    dti  delinq_2yrs earliest_cr_line  inq_last_6mths  \
0         FL  19.90          0.0         Apr-1995             1.0   
1         FL  20.12          0.0         Mar-2010             0.0   
2         CA  13.51          0.0         Apr-2007             0.0   
3         PA  38.09          0.0         Mar-1998             1.0   
4         GA  31.03          0.0         Sep-2006             0.0   

   mths_since_last_delinq  mths_since_last_record  open_acc  pub_rec  \
0                     NaN                     NaN      24.0      0.0   
1                     NaN                     NaN      13.0      0.0   
2                    44.0                    88.0       9.0      1.0   
3                    47.0                     NaN      14.0      0.0   
4                     NaN                     NaN       7.0      0.0   

   revol_bal revol_util  total_acc initial_list_status  out_prncp  \
0    57259.0      43.2%       51.0                   w       0.00   
1    11720.0      47.1%       26.0                   f       0.00   
2    11859.0      53.9%       11.0                   w    8337.74   
3    15323.0        53%       21.0                   w   10663.58   
4     4518.0      28.6%       11.0                   w    2378.98   

   out_prncp_inv   total_pymnt  total_pymnt_inv  total_rec_prncp  \
0           0.00  38187.046837         38187.05         35000.00   
1           0.00   5615.977674          5615.98          5000.00   
2        8337.74   4527.860000          4527.86          1662.26   
3       10663.58   9993.810000          9993.81          6436.42   
4        2378.98   2073.790000          2073.79          1621.02   

   total_rec_int  total_rec_late_fee  recoveries  collection_recovery_fee  \
0        3187.05                 0.0         0.0                      0.0   
1         615.98                 0.0         0.0                      0.0   
2        2865.60                 0.0         0.0                      0.0   
3        3557.39                 0.0         0.0                      0.0   
4         452.77                 0.0         0.0                      0.0   

  last_pymnt_d  last_pymnt_amnt nex

Lets look at the NaN values of each column so that you can see the problem that the extra rows at the bottom of the file are creating for us

In [0]:
# Sum null values by column and sort from least to greatest
df.isnull().sum().sort_values()

inq_fi                                             2
mo_sin_old_rev_tl_op                               2
delinq_amnt                                        2
chargeoff_within_12_mths                           2
acc_open_past_24mths                               2
inq_last_12m                                       2
total_cu_tl                                        2
total_rev_hi_lim                                   2
open_rv_24m                                        2
open_rv_12m                                        2
total_bal_il                                       2
open_il_24m                                        2
open_il_12m                                        2
open_act_il                                        2
open_acc_6m                                        2
tot_cur_bal                                        2
tot_coll_amt                                       2
acc_now_delinq                                     2
application_type                              

In [0]:
df.tail()

id  member_id  \
128409                                               NaN        NaN   
128410                                               NaN        NaN   
128411                                               NaN        NaN   
128412  Total amount funded in policy code 1: 2050909275        NaN   
128413   Total amount funded in policy code 2: 820109297        NaN   

        loan_amnt  funded_amnt  funded_amnt_inv        term int_rate  \
128409     5000.0       5000.0           5000.0   36 months   13.56%   
128410    10000.0      10000.0           9750.0   36 months   11.06%   
128411    10000.0      10000.0          10000.0   36 months   16.91%   
128412        NaN          NaN              NaN         NaN      NaN   
128413        NaN          NaN              NaN         NaN      NaN   

        installment grade sub_grade             emp_title emp_length  \
128409       169.83     C        C1          Payoff Clerk  10+ years   
128410       327.68     B        B3                   NaN        NaN   
128411       356.08     C        C5  Key Accounts Manager    2 years   
128412          NaN   NaN       NaN                   NaN        NaN   
128413          NaN   NaN       NaN                   NaN        NaN   

       home_ownership  annual_inc verification_status   issue_d loan_status  \
128409       MORTGAGE     35360.0        Not Verified  Oct-2018     Current   
128410           RENT     44400.0     Source Verified  Oct-2018     Current   
128411           RENT     80000.0        Not Verified  Oct-2018     Current   
128412            NaN         NaN                 NaN       NaN         NaN   
128413            NaN         NaN                 NaN       NaN         NaN   

       pymnt_plan  url  desc             purpose                    title  \
128409          n  NaN   NaN  debt_consolidation       Debt consolidation   
128410          n  NaN   NaN         credit_card  Credit card refinancing   
128411          n  NaN   NaN               other                    Other   
128412        NaN  NaN   NaN                 NaN                      NaN   
128413        NaN  NaN   NaN                 NaN                      NaN   

       zip_code addr_state    dti  delinq_2yrs earliest_cr_line  \
128409    381xx         TN  11.30          1.0         Jun-2006   
128410    980xx         WA  11.78          0.0         Oct-2008   
128411    021xx         MA  17.72          1.0         Sep-2006   
128412      NaN        NaN    NaN          NaN              NaN   
128413      NaN        NaN    NaN          NaN              NaN   

        inq_last_6mths  mths_since_last_delinq  mths_since_last_record  \
128409             0.0                    21.0                     NaN   
128410             2.0                    40.0                     NaN   
128411             0.0                    14.0                     NaN   
128412             NaN                     NaN                     NaN   
128413             NaN                     NaN                     NaN   

        open_acc  pub_rec  revol_bal revol_util  total_acc  \
128409       9.0      0.0     2597.0      27.3%       15.0   
128410      15.0      0.0     6269.0      13.1%       25.0   
128411      17.0      0.0     1942.0      30.8%       31.0   
128412       NaN      NaN        NaN        NaN        NaN   
128413       NaN      NaN        NaN        NaN        NaN   

       initial_list_status  out_prncp  out_prncp_inv  total_pymnt  \
128409                   f    2751.79        2751.79      3053.17   
128410                   f    5411.74        5276.45      5882.88   
128411                   w    5626.39        5626.39      6400.05   
128412                 NaN        NaN            NaN          NaN   
128413                 NaN        NaN            NaN          NaN   

        total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
128409          3053.17          2248.21         804.96                 0.0   
128410          5735.81      

In [0]:
# Address the extra NaNs in each column by skipping the footer as well.
df = pd.read_csv('LoanStats_2018Q4.csv', header=1, skipfooter=2, engine="python")

print(df.shape)
df.head()

(128412, 144)


id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0 NaN        NaN      35000        35000          35000.0   36 months   
1 NaN        NaN       5000         5000           5000.0   36 months   
2 NaN        NaN      10000        10000          10000.0   60 months   
3 NaN        NaN      17100        17100          17100.0   36 months   
4 NaN        NaN       4000         4000           4000.0   36 months   

  int_rate  installment grade sub_grade             emp_title emp_length  \
0   14.47%      1204.23     C        C2       Staff Physician    8 years   
1   22.35%       191.86     D        D5     Director of Sales  10+ years   
2   23.40%       284.21     E        E1                   NaN   < 1 year   
3   18.94%       626.30     D        D2         Receptionist   10+ years   
4   10.72%       130.43     B        B2  Extrusion assistant   10+ years   

  home_ownership  annual_inc verification_status   issue_d loan_status  \
0       MORTGAGE    360000.0            Verified  Dec-2018  Fully Paid   
1       MORTGAGE     72000.0     Source Verified  Dec-2018  Fully Paid   
2           RENT     55000.0     Source Verified  Dec-2018     Current   
3           RENT     38000.0            Verified  Dec-2018     Current   
4       MORTGAGE     56000.0        Not Verified  Dec-2018     Current   

  pymnt_plan  url  desc             purpose                    title zip_code  \
0          n  NaN   NaN         credit_card  Credit card refinancing    336xx   
1          n  NaN   NaN  debt_consolidation       Debt consolidation    333xx   
2          n  NaN   NaN  debt_consolidation       Debt consolidation    902xx   
3          n  NaN   NaN  debt_consolidation       Debt consolidation    150xx   
4          n  NaN   NaN         credit_card  Credit card refinancing    301xx   

  addr_state    dti  delinq_2yrs earliest_cr_line  inq_last_6mths  \
0         FL  19.90            0         Apr-1995               1   
1         FL  20.12            0         Mar-2010               0   
2         CA  13.51            0         Apr-2007               0   
3         PA  38.09            0         Mar-1998               1   
4         GA  31.03            0         Sep-2006               0   

   mths_since_last_delinq  mths_since_last_record  open_acc  pub_rec  \
0                     NaN                     NaN        24        0   
1                     NaN                     NaN        13        0   
2                    44.0                    88.0         9        1   
3                    47.0                     NaN        14        0   
4                     NaN                     NaN         7        0   

   revol_bal revol_util  total_acc initial_list_status  out_prncp  \
0      57259      43.2%         51                   w       0.00   
1      11720      47.1%         26                   f       0.00   
2      11859      53.9%         11                   w    8337.74   
3      15323        53%         21                   w   10663.58   
4       4518      28.6%         11                   w    2378.98   

   out_prncp_inv   total_pymnt  total_pymnt_inv  total_rec_prncp  \
0           0.00  38187.046837         38187.05         35000.00   
1           0.00   5615.977674          5615.98          5000.00   
2        8337.74   4527.860000          4527.86          1662.26   
3       10663.58   9993.810000          9993.81          6436.42   
4        2378.98   2073.790000          2073.79          1621.02   

   total_rec_int  total_rec_late_fee  recoveries  collection_recovery_fee  \
0        3187.05                 0.0         0.0                      0.0   
1         615.98                 0.0         0.0                      0.0   
2        2865.60                 0.0         0.0                      0.0   
3        3557.39                 0.0         0.0                      0.0   
4         452.77                 0.0         0.0                      0.0   

  last_pymnt_d  last_pymnt_amnt next_pym

For good measure, we'll also drop some columns that are made up completely of NaN values.

Why might LendingClub have included columns in their dataset that are 100% blank?

In [0]:
df = df.drop(['url', 'member_id', 'id', 'desc'], axis=1)
#does not run

In [0]:
df.head()

loan_amnt  funded_amnt  funded_amnt_inv        term int_rate  installment  \
0      35000        35000          35000.0   36 months   14.47%      1204.23   
1       5000         5000           5000.0   36 months   22.35%       191.86   
2      10000        10000          10000.0   60 months   23.40%       284.21   
3      17100        17100          17100.0   36 months   18.94%       626.30   
4       4000         4000           4000.0   36 months   10.72%       130.43   

  grade sub_grade             emp_title emp_length home_ownership  annual_inc  \
0     C        C2       Staff Physician    8 years       MORTGAGE    360000.0   
1     D        D5     Director of Sales  10+ years       MORTGAGE     72000.0   
2     E        E1                   NaN   < 1 year           RENT     55000.0   
3     D        D2         Receptionist   10+ years           RENT     38000.0   
4     B        B2  Extrusion assistant   10+ years       MORTGAGE     56000.0   

  verification_status   issue_d loan_status pymnt_plan             purpose  \
0            Verified  Dec-2018  Fully Paid          n         credit_card   
1     Source Verified  Dec-2018  Fully Paid          n  debt_consolidation   
2     Source Verified  Dec-2018     Current          n  debt_consolidation   
3            Verified  Dec-2018     Current          n  debt_consolidation   
4        Not Verified  Dec-2018     Current          n         credit_card   

                     title zip_code addr_state    dti  delinq_2yrs  \
0  Credit card refinancing    336xx         FL  19.90            0   
1       Debt consolidation    333xx         FL  20.12            0   
2       Debt consolidation    902xx         CA  13.51            0   
3       Debt consolidation    150xx         PA  38.09            0   
4  Credit card refinancing    301xx         GA  31.03            0   

  earliest_cr_line  inq_last_6mths  mths_since_last_delinq  \
0         Apr-1995               1                     NaN   
1         Mar-2010               0                     NaN   
2         Apr-2007               0                    44.0   
3         Mar-1998               1                    47.0   
4         Sep-2006               0                     NaN   

   mths_since_last_record  open_acc  pub_rec  revol_bal revol_util  total_acc  \
0                     NaN        24        0      57259      43.2%         51   
1                     NaN        13        0      11720      47.1%         26   
2                    88.0         9        1      11859      53.9%         11   
3                     NaN        14        0      15323        53%         21   
4                     NaN         7        0       4518      28.6%         11   

  initial_list_status  out_prncp  out_prncp_inv   total_pymnt  \
0                   w       0.00           0.00  38187.046837   
1                   f       0.00           0.00   5615.977674   
2                   w    8337.74        8337.74   4527.860000   
3                   w   10663.58       10663.58   9993.810000   
4                   w    2378.98        2378.98   2073.790000   

   total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
0         38187.05         35000.00        3187.05                 0.0   
1          5615.98          5000.00         615.98                 0.0   
2          4527.86          1662.26        2865.60                 0.0   
3          9993.81          6436.42        3557.39                 0.0   
4          2073.79          1621.02         452.77                 0.0   

   recoveries  collection_recovery_fee last_pymnt_d  last_pymnt_amnt  \
0         0.0                      0.0     Aug-2019         29882.16   
1         0.0                      0.0     Jul-2019          4474.13   
2         0.0                      0.0     Apr-2020           284.21   
3         0.0                      0.0     Apr-2020           626.30   
4         0.0                      0.0     Apr-2020           130.43   

  next_pymnt_d l

### Clean up the `int_rate` column

When we're preparing a dataset for a machine learning model we typically want to represent don't want to leave any string values in our dataset --because it's hard to do math on words. 

Specifically, we have a column that is representing a numeric value, but currently doesn't have a numeric datatype. Lets look at the first 10 values of the `int_rate` column

In [0]:
# Look at the first 10 values of the int_rate column
df['int_rate'].head(10)

0     14.47%
1     22.35%
2     23.40%
3     18.94%
4     10.72%
5     11.31%
6     11.31%
7     11.80%
8     11.80%
9     18.94%
Name: int_rate, dtype: object

In [0]:
# Look at a specific value from the int_rate column
df['int_rate'][3]

' 18.94%'

Problems that we need to address with this column:

- String column that should be numeric
- Percent Sign `%` included with the number
- Leading space at the beginning of the string

However, we're not going to try and write exactly the right code to fix this column in one go. We're going to methodically build up to the code that will help us address these problems.


In [0]:
# Lets start with just fixing a single string.
# If we can fix one, we can usually fix all of them
int_rate = ' 18.94%'

In [0]:
int_rate.strip('%')

' 18.94'

In [0]:
int_rate.strip()

'18.94%'

In [0]:
int_rate.strip('%').strip()

'18.94'

In [0]:
type(int_rate.strip('%').strip())

str

In [0]:
# "Cast" the string value to a float
float(int_rate.strip('%').strip())

18.94

In [0]:
type(float(int_rate.strip('%').strip()))

float

### Write a function to make our solution reusable!

In [0]:
# Write a function that can do what we have written above to any 
# string that is passsed to it.
def int_rate_to_float(int_rate):
  return float(int_rate.strip('%').strip())


In [0]:
# Test out our function by calling it on our example
int_rate_to_float(' 18.94%')

18.94

In [0]:
# is the data type correct?
type(int_rate_to_float(' 18.94%'))

### Apply our solution to every cell in a column

In [0]:
df['int_rate'] = df['int_rate'].apply(int_rate_to_float)

df.head()

loan_amnt  funded_amnt  funded_amnt_inv        term  int_rate  installment  \
0      35000        35000          35000.0   36 months     14.47      1204.23   
1       5000         5000           5000.0   36 months     22.35       191.86   
2      10000        10000          10000.0   60 months     23.40       284.21   
3      17100        17100          17100.0   36 months     18.94       626.30   
4       4000         4000           4000.0   36 months     10.72       130.43   

  grade sub_grade             emp_title emp_length home_ownership  annual_inc  \
0     C        C2       Staff Physician    8 years       MORTGAGE    360000.0   
1     D        D5     Director of Sales  10+ years       MORTGAGE     72000.0   
2     E        E1                   NaN   < 1 year           RENT     55000.0   
3     D        D2         Receptionist   10+ years           RENT     38000.0   
4     B        B2  Extrusion assistant   10+ years       MORTGAGE     56000.0   

  verification_status   issue_d loan_status pymnt_plan             purpose  \
0            Verified  Dec-2018  Fully Paid          n         credit_card   
1     Source Verified  Dec-2018  Fully Paid          n  debt_consolidation   
2     Source Verified  Dec-2018     Current          n  debt_consolidation   
3            Verified  Dec-2018     Current          n  debt_consolidation   
4        Not Verified  Dec-2018     Current          n         credit_card   

                     title zip_code addr_state    dti  delinq_2yrs  \
0  Credit card refinancing    336xx         FL  19.90            0   
1       Debt consolidation    333xx         FL  20.12            0   
2       Debt consolidation    902xx         CA  13.51            0   
3       Debt consolidation    150xx         PA  38.09            0   
4  Credit card refinancing    301xx         GA  31.03            0   

  earliest_cr_line  inq_last_6mths  mths_since_last_delinq  \
0         Apr-1995               1                     NaN   
1         Mar-2010               0                     NaN   
2         Apr-2007               0                    44.0   
3         Mar-1998               1                    47.0   
4         Sep-2006               0                     NaN   

   mths_since_last_record  open_acc  pub_rec  revol_bal revol_util  total_acc  \
0                     NaN        24        0      57259      43.2%         51   
1                     NaN        13        0      11720      47.1%         26   
2                    88.0         9        1      11859      53.9%         11   
3                     NaN        14        0      15323        53%         21   
4                     NaN         7        0       4518      28.6%         11   

  initial_list_status  out_prncp  out_prncp_inv   total_pymnt  \
0                   w       0.00           0.00  38187.046837   
1                   f       0.00           0.00   5615.977674   
2                   w    8337.74        8337.74   4527.860000   
3                   w   10663.58       10663.58   9993.810000   
4                   w    2378.98        2378.98   2073.790000   

   total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
0         38187.05         35000.00        3187.05                 0.0   
1          5615.98          5000.00         615.98                 0.0   
2          4527.86          1662.26        2865.60                 0.0   
3          9993.81          6436.42        3557.39                 0.0   
4          2073.79          1621.02         452.77                 0.0   

   recoveries  collection_recovery_fee last_pymnt_d  last_pymnt_amnt  \
0         0.0                      0.0     Aug-2019         29882.16   
1         0.0                      0.0     Jul-2019          4474.13   
2         0.0                      0.0     Apr-2020           284.21   
3         0.0                      0.0     Apr-2020           626.30   
4         0.0                      0.0     Apr-2020           130.43   

  next_pym

In [0]:
# What type of data is held in our new column?
df.dtypes
# Look at the datatypes of the last 5 columns

loan_amnt                                       int64
funded_amnt                                     int64
funded_amnt_inv                               float64
term                                           object
int_rate                                      float64
installment                                   float64
grade                                          object
sub_grade                                      object
emp_title                                      object
emp_length                                     object
home_ownership                                 object
annual_inc                                    float64
verification_status                            object
issue_d                                        object
loan_status                                    object
pymnt_plan                                     object
purpose                                        object
title                                          object
zip_code                    

## Challenge

We can create a new column with our cleaned values or overwrite the original, whatever we think best suits our needs. On your assignment you will take the same approach in trying to methodically build up the complexity of your code until you have a few lines that will work for any cell in a column. At that point you'll contain all of that functionality in a reusable function block and then use the `.apply()` function to... well... apply those changes to an entire column.

# [Objective](#pandas-apply) Modify and Create Columns using `.apply()`



## Overview

We're already seen one example of using the `.apply()` function to clean up a column. Lets see if we can do it again, but this time on a slightly more complicated use case.

Remember, the goal here is to write a function that will work correctly on any **individual** cell of a specific column. Then we can reuse that function on those individual cells of a dataframe column via the `.apply()` function.

Lets clean up the emp_title "Employment Title" column!

## Follow Along

First we'll try and diagnose how bad the problem is and what improvements we might be able to make.

In [0]:
# Look at the top 20 employment titles
df['emp_title'].value_counts().head(20)

Teacher                     2090
Manager                     1773
Registered Nurse             952
Driver                       924
RN                           726
Supervisor                   697
Sales                        580
Project Manager              526
General Manager              523
Office Manager               521
Owner                        420
Director                     402
Operations Manager           387
Truck Driver                 387
Nurse                        326
Engineer                     325
Sales Manager                304
manager                      301
Supervisor                   270
Administrative Assistant     269
Name: emp_title, dtype: int64

In [0]:
# How many different unique employment titles are there currently?
df['emp_title'].nunique()

43892

In [79]:
# How often is the employment_title null?
df['emp_title'].value_counts(dropna=False).head(20)

NaN                   20947
Teacher                2090
Manager                1773
Registered Nurse        952
Driver                  924
RN                      726
Supervisor              697
Sales                   580
Project Manager         526
General Manager         523
Office Manager          521
Owner                   420
Director                402
Operations Manager      387
Truck Driver            387
Nurse                   326
Engineer                325
Sales Manager           304
manager                 301
Supervisor              270
Name: emp_title, dtype: int64

What are some possible reasons as to why a person's employment title may have not been provided?

In [0]:
# Create some examples that represent the cases that we want to clean up
examples = ['owner', 'SUPERVISOR', ' Project Manager', np.NaN]

In [82]:
# Write a function to clean up these use cases and increase uniformity.
def clean_emp_title(title):
  if isinstance(title, str):
    return title.strip().title()
  else:
    return "Unknown"

for example in examples:
  print(clean_emp_title(example))

Owner
Supervisor
Project Manager
Unknown


In [83]:
# list comprehensions can combine function calls and for loops over lists
# into one succinct and fairly readable single line of code.
[clean_emp_title(title) for title in examples]

['Owner', 'Supervisor', 'Project Manager', 'Unknown']

In [0]:
# We have a function that works as expected. Lets apply it to our column.
# This time we'll overwrite the original column
df['emp_title'] = df['emp_title'].apply(clean_emp_title)

We can use the same code as we did earlier to see how much progress was made.


In [86]:
# Look at the top 20 employment titles
df['emp_title'].value_counts(dropna=False).head(20)

Unknown               20947
Teacher                2557
Manager                2395
Registered Nurse       1418
Driver                 1258
Supervisor             1160
Truck Driver            920
Rn                      834
Office Manager          805
Sales                   803
General Manager         791
Project Manager         720
Owner                   625
Director                523
Operations Manager      518
Sales Manager           500
Police Officer          440
Nurse                   425
Technician              420
Engineer                412
Name: emp_title, dtype: int64

In [87]:
# How many different unique employment titles are there currently?
df['emp_title'].nunique()

34902

In [88]:
# How often is the employment_title null (NaN)?
df['emp_title'].isnull().sum()

0

## Challenge

Using the .apply() function isn't always about creating new columns on a dataframe, we can use it to clean up or modify existing columns as well. 

# [Objective](#dates-and-times) Work with Dates and Times with Pandas

## Overview

Pandas has its own datatype datatype that makes it extremely convenient to convert strings that are in standard date formates to datetime objects and then use those datetime objects to either create new features on a dataframe or work with the dataset in a timeseries fashion. 

This section will demonstrate how to take a column of date strings, convert it to a datetime object and then use the datetime formatting `.dt` to access specific parts of the date (year, month, day) to generate useful columns on a dataframe.

## Follow Along

### Work with Dates 

pandas documentation
- [to_datetime](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.to_datetime.html)
- [Time/Date Components](https://pandas.pydata.org/pandas-docs/stable/timeseries.html#time-date-components) "You can access these properties via the `.dt` accessor"

Many of the most useful date columns in this dataset have the suffix `_d` to indicate that they correspond to dates.

We'll use a list comprehension to print them out

In [89]:
[col for col in df if col.endswith('_d')]

['issue_d', 'last_pymnt_d', 'next_pymnt_d', 'last_credit_pull_d']

Lets look at the string format of the `issue_d` column

In [90]:
df['issue_d'].head()

0    Dec-2018
1    Dec-2018
2    Dec-2018
3    Dec-2018
4    Dec-2018
Name: issue_d, dtype: object

In [92]:
df['issue_d'].value_counts()

Oct-2018    46305
Nov-2018    41973
Dec-2018    40134
Name: issue_d, dtype: int64

Because this string format %m-%y is a common datetime format, we can just let Pandas detect this format and translate it to the appropriate datetime object.

In [0]:
df['issue_d'] = pd.to_datetime(df['issue_d'], infer_datetime_format=True)

Now we can see that the `issue_d` column has been changed to hold `datetime` objects.

Lets look at one of the cells specifically to see what a datetime object looks like:

In [94]:
df['issue_d'][0]

Timestamp('2018-12-01 00:00:00')

You can see how the month and year have been indicated by the strings that were contained in the column previously, and that the rest of the values have been inferred.

In [95]:
df['issue_d'].head()

0   2018-12-01
1   2018-12-01
2   2018-12-01
3   2018-12-01
4   2018-12-01
Name: issue_d, dtype: datetime64[ns]

We can use the `.dt` accessor to now grab specific parts of the datetime object. Lets grab just the year from the all of the cells in the `issue_d` column

In [96]:
df['issue_d'].dt.year

0         2018
1         2018
2         2018
3         2018
4         2018
          ... 
128407    2018
128408    2018
128409    2018
128410    2018
128411    2018
Name: issue_d, Length: 128412, dtype: int64

Now the month.

In [97]:
df['issue_d'].dt.month

0         12
1         12
2         12
3         12
4         12
          ..
128407    10
128408    10
128409    10
128410    10
128411    10
Name: issue_d, Length: 128412, dtype: int64

It's just that easy! Now, instead of printing them out, lets add these year and month values as new columns on our dataframe. Again, you'll have to scroll all the way over to the right in the table to see the new columns.

In [98]:
df['issue_year'] = df['issue_d'].dt.year
df['issue_month'] = df['issue_d'].dt.month

df.head()

loan_amnt  funded_amnt  funded_amnt_inv        term  int_rate  installment  \
0      35000        35000          35000.0   36 months     14.47      1204.23   
1       5000         5000           5000.0   36 months     22.35       191.86   
2      10000        10000          10000.0   60 months     23.40       284.21   
3      17100        17100          17100.0   36 months     18.94       626.30   
4       4000         4000           4000.0   36 months     10.72       130.43   

  grade sub_grade            emp_title emp_length home_ownership  annual_inc  \
0     C        C2      Staff Physician    8 years       MORTGAGE    360000.0   
1     D        D5    Director Of Sales  10+ years       MORTGAGE     72000.0   
2     E        E1              Unknown   < 1 year           RENT     55000.0   
3     D        D2         Receptionist  10+ years           RENT     38000.0   
4     B        B2  Extrusion Assistant  10+ years       MORTGAGE     56000.0   

  verification_status    issue_d loan_status pymnt_plan             purpose  \
0            Verified 2018-12-01  Fully Paid          n         credit_card   
1     Source Verified 2018-12-01  Fully Paid          n  debt_consolidation   
2     Source Verified 2018-12-01     Current          n  debt_consolidation   
3            Verified 2018-12-01     Current          n  debt_consolidation   
4        Not Verified 2018-12-01     Current          n         credit_card   

                     title zip_code addr_state    dti  delinq_2yrs  \
0  Credit card refinancing    336xx         FL  19.90            0   
1       Debt consolidation    333xx         FL  20.12            0   
2       Debt consolidation    902xx         CA  13.51            0   
3       Debt consolidation    150xx         PA  38.09            0   
4  Credit card refinancing    301xx         GA  31.03            0   

  earliest_cr_line  inq_last_6mths  mths_since_last_delinq  \
0         Apr-1995               1                     NaN   
1         Mar-2010               0                     NaN   
2         Apr-2007               0                    44.0   
3         Mar-1998               1                    47.0   
4         Sep-2006               0                     NaN   

   mths_since_last_record  open_acc  pub_rec  revol_bal revol_util  total_acc  \
0                     NaN        24        0      57259      43.2%         51   
1                     NaN        13        0      11720      47.1%         26   
2                    88.0         9        1      11859      53.9%         11   
3                     NaN        14        0      15323        53%         21   
4                     NaN         7        0       4518      28.6%         11   

  initial_list_status  out_prncp  out_prncp_inv   total_pymnt  \
0                   w       0.00           0.00  38187.046837   
1                   f       0.00           0.00   5615.977674   
2                   w    8337.74        8337.74   4527.860000   
3                   w   10663.58       10663.58   9993.810000   
4                   w    2378.98        2378.98   2073.790000   

   total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
0         38187.05         35000.00        3187.05                 0.0   
1          5615.98          5000.00         615.98                 0.0   
2          4527.86          1662.26        2865.60                 0.0   
3          9993.81          6436.42        3557.39                 0.0   
4          2073.79          1621.02         452.77                 0.0   

   recoveries  collection_recovery_fee last_pymnt_d  last_pymnt_amnt  \
0         0.0                      0.0     Aug-2019         29882.16   
1         0.0                      0.0     Jul-2019          4474.13   
2         0.0                      0.0     Apr-2020           284.21   
3         0.0                      0.0     Apr-2020           626.30   
4         0.0                      0.0     Apr-2020           130.43   

  next_pym

In [99]:
df.dtypes[-5:]

settlement_amount        float64
settlement_percentage    float64
settlement_term          float64
issue_year                 int64
issue_month                int64
dtype: object

Because all of these dates come from Q4 of 2018, the `issue_d` column isn't all that interesting. Lets look at the `earliest_cr_line` column, which is also a string, but that could be converted to datetime format.

We're going to create a new column called `days_from_earliest_credit_to_issue`

It's a long column header, but think about how valuable this piece of information could be. This number will essentially indicate the length of a person's credit history and if that is correlated with repayment or other factors could be a valuable predictor!

In [0]:
df['earliest_cr_line'] = pd.to_datetime(df['earliest_cr_line'], infer_datetime_format=True)

In [101]:
df['earliest_cr_line'].head()

0   1995-04-01
1   2010-03-01
2   2007-04-01
3   1998-03-01
4   2006-09-01
Name: earliest_cr_line, dtype: datetime64[ns]

What we're about to do is so cool! Pandas' datetime format is so smart that we can simply use the subtraction operator `-` in order to calculate the amount of time between two dates. 

Think about everything that's going on under the hood in order to give us such straightforward syntax! Handling months of different lengths, leap years, etc. Pandas datetime objects are seriously powerful!

In [0]:
df['years_from_earliest_credit_to_issue'] = ((df['issue_d'] - df['earliest_cr_line']).dt.days / 365)

In [103]:
df.head()

loan_amnt  funded_amnt  funded_amnt_inv        term  int_rate  installment  \
0      35000        35000          35000.0   36 months     14.47      1204.23   
1       5000         5000           5000.0   36 months     22.35       191.86   
2      10000        10000          10000.0   60 months     23.40       284.21   
3      17100        17100          17100.0   36 months     18.94       626.30   
4       4000         4000           4000.0   36 months     10.72       130.43   

  grade sub_grade            emp_title emp_length home_ownership  annual_inc  \
0     C        C2      Staff Physician    8 years       MORTGAGE    360000.0   
1     D        D5    Director Of Sales  10+ years       MORTGAGE     72000.0   
2     E        E1              Unknown   < 1 year           RENT     55000.0   
3     D        D2         Receptionist  10+ years           RENT     38000.0   
4     B        B2  Extrusion Assistant  10+ years       MORTGAGE     56000.0   

  verification_status    issue_d loan_status pymnt_plan             purpose  \
0            Verified 2018-12-01  Fully Paid          n         credit_card   
1     Source Verified 2018-12-01  Fully Paid          n  debt_consolidation   
2     Source Verified 2018-12-01     Current          n  debt_consolidation   
3            Verified 2018-12-01     Current          n  debt_consolidation   
4        Not Verified 2018-12-01     Current          n         credit_card   

                     title zip_code addr_state    dti  delinq_2yrs  \
0  Credit card refinancing    336xx         FL  19.90            0   
1       Debt consolidation    333xx         FL  20.12            0   
2       Debt consolidation    902xx         CA  13.51            0   
3       Debt consolidation    150xx         PA  38.09            0   
4  Credit card refinancing    301xx         GA  31.03            0   

  earliest_cr_line  inq_last_6mths  mths_since_last_delinq  \
0       1995-04-01               1                     NaN   
1       2010-03-01               0                     NaN   
2       2007-04-01               0                    44.0   
3       1998-03-01               1                    47.0   
4       2006-09-01               0                     NaN   

   mths_since_last_record  open_acc  pub_rec  revol_bal revol_util  total_acc  \
0                     NaN        24        0      57259      43.2%         51   
1                     NaN        13        0      11720      47.1%         26   
2                    88.0         9        1      11859      53.9%         11   
3                     NaN        14        0      15323        53%         21   
4                     NaN         7        0       4518      28.6%         11   

  initial_list_status  out_prncp  out_prncp_inv   total_pymnt  \
0                   w       0.00           0.00  38187.046837   
1                   f       0.00           0.00   5615.977674   
2                   w    8337.74        8337.74   4527.860000   
3                   w   10663.58       10663.58   9993.810000   
4                   w    2378.98        2378.98   2073.790000   

   total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
0         38187.05         35000.00        3187.05                 0.0   
1          5615.98          5000.00         615.98                 0.0   
2          4527.86          1662.26        2865.60                 0.0   
3          9993.81          6436.42        3557.39                 0.0   
4          2073.79          1621.02         452.77                 0.0   

   recoveries  collection_recovery_fee last_pymnt_d  last_pymnt_amnt  \
0         0.0                      0.0     Aug-2019         29882.16   
1         0.0                      0.0     Jul-2019          4474.13   
2         0.0                      0.0     Apr-2020           284.21   
3         0.0                      0.0     Apr-2020           626.30   
4         0.0                      0.0     Apr-2020           130.43   

  next_pym

What's oldest credit history that was involved in Q4 2018? 

In [104]:
df['years_from_earliest_credit_to_issue'].describe()

count    128412.000000
mean         16.054497
std           7.908317
min           3.084932
25%          11.093151
50%          14.427397
75%          19.846575
max          68.961644
Name: years_from_earliest_credit_to_issue, dtype: float64

25,171 days is ~ 68.96 years of credit history!

## Challenge

Pandas' datetime format is so easy to work with that there's really no excuse for not using dates to make features on a dataframe! Get ready to practice more of this on your assignment.